src/coreclr/jit/valuenum.cpp

```diff
@@ -8492,7 +8492,98 @@ void Compiler::fgValueNumberSsaVarDef(GenTreeLclVarCommon* lcl)
+    {
+        return false;
+    }
+
+    // Is given VN representing a frozen object handle
+    auto isCnsObjHandle = [](ValueNumStore* vnStore, ValueNum vn, CORINFO_OBJECT_HANDLE* handle) -> bool {
+        if (vnStore->IsVNHandle(vn) && (vnStore->GetHandleFlags(vn) == GTF_ICON_OBJ_HDL))
+        {
+            const size_t obj = vnStore->CoercedConstantValue<size_t>(vn);
+            *handle          = reinterpret_cast<CORINFO_OBJECT_HANDLE>(obj);
+            return true;
+        }
+        return false;
+    };
+
+    CORINFO_OBJECT_HANDLE objHandle = nullptr;
+    int                   index     = -1;
+
+    // First, let see if we have PtrToArrElem
+    ValueNum addr = funcApp.m_args[0];
+    if (funcApp.m_func == VNF_PtrToArrElem)
+    {
+        ValueNum arrVN  = funcApp.m_args[1];
+        ValueNum inxVN  = funcApp.m_args[2];
+        ssize_t  offset = vnStore->ConstantValue<ssize_t>(funcApp.m_args[3]);
+
+        if (isCnsObjHandle(vnStore, arrVN, &objHandle) && offset == 0 && vnStore->IsVNConstant(inxVN))
+        {
+            index = (int)vnStore->CoercedConstantValue<size_t>(inxVN);
+        }
+    }
+    else if (((genTreeOps)funcApp.m_func == GT_ADD) && vnStore->IsVNConstant(funcApp.m_args[0]))
+    {
+        // Here we try to match these two patterns:
+        //
+        //  1) (objHandle + firstCharOffset)
+        //  2) ((objHandle + firstCharOffset) + byteIndex)
+        //
+        size_t dataOffset = 0;
+        if (vnStore->IsVNConstant(funcApp.m_args[1]) && isCnsObjHandle(vnStore, funcApp.m_args[0], &objHandle))
+        {
+            dataOffset = vnStore->CoercedConstantValue<size_t>(funcApp.m_args[1]);
+        }
+        else
+        {
+            assert(vnStore->IsVNConstant(funcApp.m_args[0]));
+            const size_t byteDataIndex = vnStore->CoercedConstantValue<size_t>(funcApp.m_args[0]);
+            if (vnStore->GetVNFunc(funcApp.m_args[1], &funcApp) && ((genTreeOps)funcApp.m_func == GT_ADD) &&
+                vnStore->IsVNConstant(funcApp.m_args[1]) && isCnsObjHandle(vnStore, funcApp.m_args[0], &objHandle))
+            {
+                dataOffset = vnStore->CoercedConstantValue<size_t>(funcApp.m_args[1]) + byteDataIndex;
+            }
+        }
+
+        // Convert dataOffset (that also includes offset to firstChar field) into char index
+        if ((dataOffset >= OFFSETOF__CORINFO_String__chars) && ((dataOffset % 2) == 0))
+        {
+            index = (int)(dataOffset - OFFSETOF__CORINFO_String__chars) / 2;
+        }
+    }
+
+    if ((objHandle != nullptr) && (index >= 0))
+    {
+        assert(objHandle != nullptr);
+        USHORT charValue;
+        if (info.compCompHnd->getStringChar(objHandle, index, &charValue))
+        {
+            JITDUMP("Folding \"cns_str\"[%d] into %u\n", index, (unsigned)charValue);
+            // NOTE: we need to sign-extend it here in case if it's a negative Int16
+            tree->gtVNPair.SetBoth(vnStore->VNForIntCon((SHORT)charValue));
```

Comment:
This assumes the load doesn't have a significant difference in the conservative component, which does not like an assumption we can make.

E. g.:
```cs
obj.StringField = "my constant";
/* here thread B mutates this to be "not my constant" */
Use(obj.StringField[2]);
```
So this optimization should either be done for `Liberal == Conservative` or (which is what all of the other similar optimizations do) for both liberal and conservative components, with some `VNForConstLoad` or similar function.

Reply:
Yes, it had `Liberal == Conservative` I probably removed it before push